modules/apps/content-dashboard/content-dashboard-web/src/main/resources/META-INF/resources/js/components/AuditBarChart.js

```diff
@@ -25,18 +25,40 @@ import {
 	YAxis,
 } from 'recharts';
 
-const BAR_CHART = {
-	barHeight: 16,
-	dotRadiusMax: 35,
-	dotRadiusMin: 5,
-	fillXAxis: '#666',
-	height: 400,
-	legendHeight: 70,
-	stroke: '#E7E7ED',
-	width: 1150,
-};
+import {BAR_CHART, COLORS} from '../utils/constants';
+
+export default function AuditBarChart({categories, rtl}) {
+	const auditBarChartData = useMemo(() => {
+		const bars = [];
+		const data = [];
+
+		categories.map((category) => {
+			var dataChild = {name: category.name};
+
+			const children = category.children;
+
+			for (var i = 0; i < children.length; i++) {
+				const barChild = {
+					dataKey: children[i].key,
+					name: children[i].name,
+				};
+				if (!bars.some((bar) => bar.dataKey === children[i].key)) {
+					bars.push(barChild);
+				}
+
+				dataChild = {
+					...dataChild,
+					[children[i].key]: children[i].value,
+				};
+			}
+			data.push(dataChild);
+		});
+
+		return {bars, data};
```

Comment:
I think we might be able to clean this up a little bit by using `.reduce` and avoid mutations if possible.

```js
const data = categories.map((category) => {
	return category.children.reduce(
		(acc, child) => {
			return {
				...acc,
				[child.key]: child.value,
			};
		},
		{name: category.name}
	);
});

const bars = categories.reduce(
	(acc, category) => {
		const dataKeys = []

		const newBar = category.children.reduce((childAcc, {key as dataKey, name}) => {
			if (dataKeys.includes(dataKey)) {
				return childAcc
			}

			dataKeys.push(dataKey)

			return [...childAcc, {dataKey, name}]
		}, [])

		return [...acc, newBar];
	},
	[]
);
```

I haven't actually tested this, but this was at least the general idea and direction.

Does that seem better? I'm not totally convinced myself, but I often find it easier to take a more functional approach when mapping through data.

Reply:
I wouldn't use an array plus `includes()` to test for membership in `dataKeys` because that requires a linear scan through the array (`O(n)`) compared with a `Set`, which would be a constant time lookup (`O(1)`). Of course, it won't have any measurable performance impact for small `n`, but we need to use the "right data structure for the job", or the likelihood that somebody will copy the pattern and end up using `includes` when `n` *is* large will go up.

```js
const dataKeys = new Set();

if (dataKeys.has(dataKey)) {
   // ...
}
```

(Same issue applied to the original version, where `bars.some` is `O(n)`.)

> Does that seem better? I'm not totally convinced myself, but I often find it easier to take a more functional approach when mapping through data.

The difference isn't large here, but in general, defaulting to a functional/immutable style is a good starting point. `reduce` and spread syntax can be hard to read at first, but you people are getting more and more used to it over time, and it makes the immutable style so much more tractable, which makes it worth it.

If you're worried about the perf of creating so many intermediate arrays, you could use `concat` instead of spread for the accumulators; eg.

```js
// instead of:
return [...acc, newBar];

// this:
return acc.concat(newBar);

// etc
```

One small syntax error:

```
key as dataKey
```

Should be:

```
key: dataKey
```

And a (possible) edit:

```js
// Instead of:
return category.children.reduce(
	(acc, child) => {
		return {
			...acc,
			[child.key]: child.value,
		};
	},
	{name: category.name}
);

// Destructuring to avoid repetition:
return category.children.reduce(
	(acc, {key, value}) => {
		return {
			...acc,
			[key]: value,
		};
	},
	{name: category.name}
);
```

Reply:
Thanks both for your feedback. I'll apply it and resent :)